Fix #71 (missing publish call in internal server subscriptions)

cbergmiller opened the pull request:
On internal server subscriptions `InternalSubscription._not_acknowledged_results` kept growing onexternal data changes. This was a bug that was introduced in the #45 Subscription refactoring.

cbergmiller commented:
This is a fix for #71 and possibly #79

oroulet commented on asyncua/server/internal_subscription.py:
```diff
@@ -97,6 +97,7 @@ def publish_results(self):
         if result is not None:
             self.logger.info('publish_results for %s', self.data.SubscriptionId)
             self.pub_result_callback(result)
+            self.publish([result.NotificationMessage.SequenceNumber])
```
Is that code only call when subscription is server side? If it is called from client too then we will receive dobble publish....

cbergmiller replied:
It _should_ be the same way it used to be before #45.
`InternalSubscription.publish` does not do much:
```python
def publish(self, acks: Iterable[int]):
        """
        Reset publish cycle count, acknowledge PublishResults.
        :param acks: Sequence number of the PublishResults to acknowledge
        """
        self.logger.info("publish request with acks %s", acks)
        self._publish_cycles_count = 0
        for nb in acks:
            self._not_acknowledged_results.pop(nb, None)
```
The `pop` already had the `None` default argument - in case the result was already acked.
I will check the client subscription too. If possible i will add a test case to prevent regression bugs in the future.

cbergmiller commented:
I removed the `publish` call from the `InternalSubscription.publish_results` again since it would break the ack logic of the PublishResults that are sent from Servers to Clients.
I did not want to go back to sending Publish Acks in `Subscription.publish_callback`. Instead i added a flag to `InternalSubscription` that is only set if the Subscription was created for a Client on the Server.
If the Subscription is internal to the Server the Publish Results are now no longer stored for republishing.

oroulet commented on asyncua/server/internal_session.py:
```diff
@@ -118,8 +118,8 @@ def call(self, params):
-    async def create_subscription(self, params, callback=None):
-        result = await self.subscription_service.create_subscription(params, callback)
+    async def create_subscription(self, params, callback=None, is_for_client=False):
+        result = await self.subscription_service.create_subscription(params, callback, is_for_client)
```
in fact the InternalSession object already has a `self.external` property which contains that information

cbergmiller replied:
Oops, thanks i will look into that!

oroulet commented on asyncua/server/internal_subscription.py:
```diff
@@ -92,10 +93,14 @@ def publish_results(self):
         if result is not None:
             self.logger.info('publish_results for %s', self.data.SubscriptionId)
+            # The callback can be:
+            #    Subscription.publish_callback -> server internal subscription
```
Subscription is also used on client side. So you need to document that Subscription.publish_callback is used for server side internal subcriptions AND client side

oroulet replied:
OK wrong comment. This was a differnt file..

oroulet commented on asyncua/server/internal_subscription.py:
```diff
@@ -17,7 +17,7 @@ class InternalSubscription:
 
     def __init__(self, loop: asyncio.AbstractEventLoop, data: ua.CreateSubscriptionResult, aspace: AddressSpace,
-                 callback=None):
+                 callback=None, is_for_client=False):
```
instead of calling that parameter is_for_client. it is probably better to explicitly state what is changes: no_acks or do_not_expect_acks

oroulet commented on asyncua/server/internal_subscription.py:
```diff
@@ -92,10 +93,14 @@ def publish_results(self):
+            # The callback can be:
+            #    Subscription.publish_callback -> server internal subscription
+            #    UaProcessor.forward_publish_response -> client subscription
             self.pub_result_callback(result)
```
you could call self.publish(result.Acknowledgements) here instead of disabling sending them. Not sure what is best....

cbergmiller replied:
I think this would break the ability to republish not acknowledged results (if so requested by the client). For server internal subscriptions the acknowledgement (and republishing) is not needed and can be disabled (?). I will try to do more testing to be sure.

oroulet replied:
yes should be fine to disable it for internal session

oroulet commented:
We can merge, can't we?

@dannmartens Could you try that branch and tell us if it changes anything?

dannmartens commented:
@oroulet I will try it out!

dannmartens commented:
@oroulet I am going to use the branch 'fix_71' and deploy again, so the test server can run over the week-end.

dannmartens commented:
Okay, the burn-in test server has been up for more than 48 hours now, receiving writes from a subscribed client without any significant change in the memory consumption. There is just a normal small undulation of allocation and garbage collection, order of magnitude 1 MB to 10 MB.

👌

oroulet commented:
so maybe that was the only memeroy leak :tada:
